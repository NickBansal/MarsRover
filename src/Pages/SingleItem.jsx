import React, { Component } from 'react'
import * as api from '../api';
import '../Stylesheets/SingleItem.css'
import Loading from '../Components/Loading'
import moment from 'moment'
import { navigate } from '@reach/router'

class SingleItem extends Component {

    state = {
        itemData: [],
        error: false,
        loading: true,
        assets: []
    }

    render() {
        const { loading, itemData, assets, error } = this.state
        return (
            <div>
                {loading && <Loading />}

                {!loading && itemData.length === 0 &&
                    <div>
                        <p className="ErrorLoad">There was a problem retreiving the information, please try another search</p>
                        <button onClick={() => navigate('/search')}>
                            <i className="fas fa-chevron-circle-left fa-5x"></i>
                        </button>
                    </div>
                }

                {error &&
                    <div>
                        <p className="ErrorLoad">Request failed, please try another search</p>
                        <button onClick={() => navigate('/search')}><i className="fas fa-chevron-circle-left fa-5x"></i></button>
                    </div>
                }

                {!loading && itemData.length > 0 &&
                    <div className="WholeSingleItemPage">
                        {itemData[0].data[0].media_type === 'image' && 
                        <div className="hvrbox">
                            <img
                                src={itemData[0].links[0].href}
                                alt={itemData[0].data[0].title}
                                className="hvrbox-layer_bottom" />
                            <div className="hvrbox-layer_top hvrbox-layer_scale">
                                <div className="hvrbox-text">
                                    <h1>{itemData[0].data[0].title.replace(new RegExp("\\-|_", "g"), ' ')}</h1>
                                    {itemData[0].data[0].description &&
                                        <p className="Description">{itemData[0].data[0].description}</p>
                                    }
                                    <p className="Created">Created: {moment(itemData[0].data[0].date_created).from()}</p>
                                </div>
                            </div>
                            <p>Please hover over the image for more detail</p>
                        </div>}
                        {itemData[0].data[0].media_type === 'video' &&
                            <div className="VideoPlayback">
                                <video width="550" height="550" controls>
                                    <source src={assets[0].href} type="video/mp4" />
                                    <source src={assets[1].href} type="video/mp4" />
                                </video>
                                <h1>{itemData[0].data[0].title.replace(new RegExp("\\-|_", "g"), ' ')}</h1>
                                {itemData[0].data[0].description &&
                                    <p className="Description">{itemData[0].data[0].description.replace(new RegExp("\\-|_", "g"), ' ')}</p>
                                }
                                <p className="Created">Created: {moment(itemData[0].data[0].date_created).from()}</p>
                            </div>
                        }
                        <button onClick={() => navigate('/search')}><i className="fas fa-chevron-circle-left fa-3x"></i></button>
                    </div>
                }
            </div>
        )
    }

    componentDidMount() {
        const getItems = api.getItems(this.props.id)
        const getAssets = api.assetData(this.props.id)
        Promise.all([getItems, getAssets])
            .then(data => {
                this.setState({
                    itemData: data[0].collection.items,
                    loading: false,
                    assets: data[1].collection.items
                })
            })
            .catch(error => {
                this.setState({
                    error: error.message
                })
            })
    }
}

export default SingleItem