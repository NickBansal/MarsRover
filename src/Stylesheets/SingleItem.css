.SingleItem {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center
}

.hvrbox,
.hvrbox * {
	-webkit-box-sizing: border-box;
	        box-sizing: border-box;
}
.hvrbox {
	position: relative;
	display: inline-block;
	overflow: hidden;
	max-width: 100%;
    height: auto;
    border-radius: 20px;
}
.hvrbox img {
    width: 550px;
    height: 550px;
    border: 5px solid white;
    border-radius: 20px;
    -webkit-box-shadow: 0px 9px 45px 0px rgba(0,0,0,0.75);
    -moz-box-shadow: 0px 9px 45px 0px rgba(0,0,0,0.75);
    box-shadow: 0px 9px 45px 0px rgba(0,0,0,0.75); 
}
.hvrbox .hvrbox-layer_bottom {
    display: block;
    border-radius: 20px;
    opacity: 0;
    animation: 1 fadein 1.5s forwards
}

@keyframes fadein {
    100% {
        opacity: 1;
    }
}

.hvrbox .hvrbox-layer_top {
	opacity: 0;
    position: absolute;
    border-radius: 20px;
	left: 0;
	right: 0;
	top: 0;
	height: 550px;
    width: 550px;
	background: rgba(0, 0, 0, 0.8);
	border: 4px solid white; 
	color: #fff;
	padding: 15px;
	-webkit-transition: all 0.4s ease-in-out 0s;
	-o-transition: all 0.4s ease-in-out 0s;
	transition: all 0.4s ease-in-out 0s;
	overflow: scroll
}
.hvrbox:hover .hvrbox-layer_top,
.hvrbox.active .hvrbox-layer_top {
	opacity: 1;
	cursor: pointer;
	padding: 20px
}
.hvrbox .hvrbox-text {
	text-align: center;
	font-size: 18px;
	display: inline-block;
}

.hvrbox .hvrbox-text_mobile {
	font-size: 15px;
	border-top: 1px solid rgb(179, 179, 179); /* for old browsers */
	border-top: 1px solid rgba(179, 179, 179, 0.7);
	margin-top: 5px;
	padding-top: 2px;
	display: none;
}
.hvrbox.active .hvrbox-text_mobile {
	display: block;
}

.hvrbox .hvrbox-layer_scale {
	border-radius: 50%;
	-webkit-transform: scale(0);
	-ms-transform: scale(0);
	transform: scale(0);
}
.hvrbox:hover .hvrbox-layer_scale,
.hvrbox.active .hvrbox-layer_scale {
	border-radius: 20px;
	-webkit-transform: scale(1);
	-ms-transform: scale(1);
	transform: scale(1);
}

.hvrbox-text h1 {
    font-size: 35px;
    margin: 0
}

.hvrbox-text p.Description {
	font-size: 20px
}

.hvrbox-text p.Created {
	font-size: 15px
}

video {
	height: 310px;
	width: 545px;
	border-radius: 20px;
	border: 3px solid white;
	background: #0000005e
}

.fa-chevron-circle-left {
	color: white;
	transition: 0.3s
}

.fa-chevron-circle-left:hover {
	cursor: pointer;
	transform: translateY(-2px);
	color: lightgray;
	transform: scale(1.2);
	transition: 0.2s
}

.fa-5x {
	font-size: 5em;
}

button {
	background: none;
	outline: none;
	border: none
}

.VideoPlayback {
	opacity: 0;
	animation: 1 fadein 1.5s forwards
}

.VideoPlayback p {
	width: 60%;
	margin: 15px auto
}

.VideoPlayback h1 {
	width: 60%;
	margin: 20px auto
}

.ErrorLoad {
	font-size: 25px
}

@media (max-width: 850px) {
	.VideoPlayback p {
		width: 80%;
	}
	.VideoPlayback h1 {
		width: 80%;
	}
}

@media (max-width: 600px) {
	.hvrbox .hvrbox-layer_top {
		height: 300px;
		width: 300px;
	}
	.hvrbox img {
		height: 300px;
		width: 300px;
	}
	.hvrbox-text h1 {
		font-size: 20px;
	}
	.hvrbox-text p.Description {
		font-size: 15px
	}
	.hvrbox-text p.Created {
		font-size: 13px
	}
	video {
		height: 175px;
		width: 300px
	}
	.VideoPlayback p {
		width: 100%;
		font-size: 15px
	}
	.VideoPlayback h1 {
		font-size: 22px
	}
	.ErrorLoad {
		font-size: 20px
	}
	.fa-5x {
		font-size: 3em;
	}
}