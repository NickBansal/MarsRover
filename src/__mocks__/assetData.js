export default [{
  "collection": {
    "version": "1.0",
    "items": [
      {
        "href": "http://images-assets.nasa.gov/image/PIA04778/PIA04778~orig.jpg"
      },
      {
        "href": "http://images-assets.nasa.gov/image/PIA04778/metadata.json"
      }
    ],
    "href": "https://images-api.nasa.gov/asset/PIA04778"
  }
}]