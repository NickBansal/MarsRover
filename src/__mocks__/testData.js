export default {
  "collection": {
    "version": "1.0",
    "links": [
      {
        "prompt": "Next",
        "rel": "next",
        "href": "https://images-api.nasa.gov/search?page=2&q=mars"
      }
    ],
    "items": [
      {
        "links": [
          {
            "rel": "preview",
            "render": "image",
            "href": "https://images-assets.nasa.gov/image/PIA04778/PIA04778~thumb.jpg"
          }
        ],
        "data": [
          {
            "nasa_id": "PIA04778",
            "date_created": "2003-10-08T21:49:42Z",
            "description_508": "Mars in Color",
            "keywords": [
              "Mars",
              "2001 Mars Odyssey"
            ],
            "title": "Mars in Color",
            "secondary_creator": "NASA/JPL/Arizona State University",
            "media_type": "image",
            "description": "Mars in Color",
            "center": "JPL"
          }
        ],
        "href": "https://images-assets.nasa.gov/image/PIA04778/collection.json"
      },
      {
        "links": [
          {
            "rel": "preview",
            "render": "image",
            "href": "https://images-assets.nasa.gov/image/PIA04947/PIA04947~thumb.jpg"
          }
        ],
        "data": [
          {
            "nasa_id": "PIA04947",
            "date_created": "2004-01-02T23:09:48Z",
            "description_508": "Dynamic Mars",
            "keywords": [
              "Mars",
              "2001 Mars Odyssey"
            ],
            "title": "Dynamic Mars",
            "secondary_creator": "NASA/JPL/ASU",
            "media_type": "image",
            "description": "Dynamic Mars",
            "center": "JPL"
          }
        ],
        "href": "https://images-assets.nasa.gov/image/PIA04947/collection.json"
      },
      {
        "links": [
          {
            "rel": "preview",
            "render": "image",
            "href": "https://images-assets.nasa.gov/image/PIA01120/PIA01120~thumb.jpg"
          }
        ],
        "data": [
          {
            "nasa_id": "PIA01120",
            "date_created": "1997-12-12T11:34:30Z",
            "description_508": "Pathfinder on Mars",
            "keywords": [
              "Mars",
              "Mars Pathfinder MPF"
            ],
            "title": "Pathfinder on Mars",
            "secondary_creator": "NASA/JPL",
            "media_type": "image",
            "description": "Pathfinder on Mars",
            "center": "JPL"
          }
        ],
        "href": "https://images-assets.nasa.gov/image/PIA01120/collection.json"
      },
      {
        "links": [
          {
            "rel": "preview",
            "render": "image",
            "href": "https://images-assets.nasa.gov/image/PIA02311/PIA02311~thumb.jpg"
          }
        ],
        "data": [
          {
            "nasa_id": "PIA02311",
            "date_created": "1999-08-25T16:00:09Z",
            "description_508": "Mars Polar Lander and Mars Pathfinder Sites Compared",
            "keywords": [
              "Mars",
              "Mars Global Surveyor MGS"
            ],
            "title": "Mars Polar Lander and Mars Pathfinder Sites Compared",
            "secondary_creator": "NASA/JPL/MSSS",
            "media_type": "image",
            "description": "Mars Polar Lander and Mars Pathfinder Sites Compared",
            "center": "JPL"
          }
        ],
        "href": "https://images-assets.nasa.gov/image/PIA02311/collection.json"
      },
      {
        "links": [
          {
            "rel": "preview",
            "render": "image",
            "href": "https://images-assets.nasa.gov/image/PIA03201/PIA03201~thumb.jpg"
          }
        ],
        "data": [
          {
            "nasa_id": "PIA03201",
            "date_created": "2005-11-15T21:13:36Z",
            "description_508": "Mars-shine",
            "keywords": [
              "Mars",
              "Mars Exploration Rover MER"
            ],
            "title": "Mars-shine",
            "secondary_creator": "NASA/JPL-Caltech/Cornell",
            "media_type": "image",
            "description": "Mars-shine",
            "center": "JPL"
          }
        ],
        "href": "https://images-assets.nasa.gov/image/PIA03201/collection.json"
      }
    ],
    "metadata": {
      "total_hits": 18600
    },
    "href": "https://images-api.nasa.gov/search?q=mars"
  }
}