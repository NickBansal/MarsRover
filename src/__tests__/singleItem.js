import React from 'react';
import SingleItem from '../Pages/SingleItem';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import '../setupTest';
import testData from "../__mocks__/testData";
import assetData from '../__mocks__/assetData'
import mockAxios from 'axios'

describe.only('<SingleItem />', () => {

    it('Component matches the snapshot', () => {
        const wrapper = mount(<SingleItem />)
        expect(toJson(wrapper)).toMatchSnapshot()
    })

    it('componentDidMount Testing', done => {
        mockAxios.get.mockImplementationOnce(() => 
            Promise.resolve({
                data: testData
            })
        );
        mockAxios.get.mockImplementationOnce(() => 
            Promise.resolve({
                data: assetData
            })
        );
        const wrapper = mount(<SingleItem />)
        
        expect(wrapper.state().loading).toBe(true)
        expect(wrapper.state().assets.length).toBe(0)
        expect(wrapper.state().itemData.length).toBe(0)
        expect(wrapper.find('Loading').exists()).toBe(true)
        console.log(wrapper.debug())

        setTimeout(() => {
            wrapper.update();

            //STATE
            expect(wrapper.state().loading).toBe(false)
            expect(wrapper.state().assets.length).toBe(2)
            expect(wrapper.state().itemData.length).toBe(5)

            // HTML PAGE ELEMENTS
            expect(wrapper.find('div.hvrbox').exists()).toBe(true)
            expect(wrapper.find('.hvrbox-text').text()).toBe('Mars in ColorMars in ColorCreated: 15 years ago')
            expect(wrapper.find('Loading').exists()).toBe(false)
            done()
        }, 50)
    })
})

