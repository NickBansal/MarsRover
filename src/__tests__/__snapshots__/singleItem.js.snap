// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleItem /> Component matches the snapshot 1`] = `
<SingleItem>
  <div>
    <Loading>
      <div
        className="wrap"
      >
        <div
          className="spinner-wrap"
        >
          <div
            className="spinner"
          >
            <i />
            <i />
            <i />
            <i />
            <i />
            <i />
            <i />
            <i />
            <i />
            <i />
            <i />
            <i />
          </div>
        </div>
      </div>
    </Loading>
  </div>
</SingleItem>
`;
